python_modules/dagster/dagster/_cli/dev.py

```diff
@@ -326,3 +346,116 @@ def _workspace_opts_to_serialized_cli_args(workspace_opts: WorkspaceOpts) -> Seq
         args.append("--use-ssl")
 
     return args
+
+
+class DagsterDevCodeServerManager:
+    """Context manager that manages the lifecycle of code servers launched by the dagster dev command."""
+
+    def __init__(
+        self,
+        instance: DagsterInstance,
+        workspace_load_target: Optional[WorkspaceLoadTarget],
+        code_server_log_level: str = "INFO",
+        server_command: GrpcServerCommand = GrpcServerCommand.API_GRPC,
+    ) -> None:
+        self._stack = ExitStack()
+
+        self._instance = check.inst_param(instance, "instance", DagsterInstance)
+        self._workspace_load_target = check.opt_inst_param(
+            workspace_load_target, "workspace_load_target", WorkspaceLoadTarget
+        )
+
+        self._grpc_server_registry: GrpcServerRegistry = self._stack.enter_context(
+            GrpcServerRegistry(
+                instance_ref=self._instance.get_ref(),
+                server_command=server_command,
+                heartbeat_ttl=WEBSERVER_GRPC_SERVER_HEARTBEAT_TTL,
+                startup_timeout=instance.code_server_process_startup_timeout,
+                log_level=code_server_log_level,
+                wait_for_processes_on_shutdown=instance.wait_for_local_code_server_processes_on_shutdown,
+                additional_timeout_msg=INCREASE_TIMEOUT_DAGSTER_YAML_MSG,
+            )
+        )
+
+        self._current_workspace: CurrentWorkspace = CurrentWorkspace(code_location_entries={})
+        # Is it ok that there's only one of these? When would this be something that needs to be called per location?
+        self.__shutdown_event = threading.Event()
+        self.__auto_restart_thread = threading.Thread(target=self.auto_restart_thread, daemon=True)
+        self.__auto_restart_thread.start()
```

Comment:
we should join this thread in the cleanup/exit function (and also join the client heartbeat thread once you add that)